### PR TITLE
add `JULIA_CONDAPKG_TOPENV`

### PR DESCRIPTION
**Pull request by t-bltg**
Possible implementation for https://github.com/cjdoris/CondaPkg.jl/issues/52.

### CODE REVIEW

**Comment by cjdoris** [comment]
Looks good so far, thanks!

On my phone right now so can't review properly but the environment written to needs to be included in the metadata, and the skip resolve check should check if the environment has changed.

**Comment by t-bltg** [comment]
Thanks,

> the environment written to needs to be included in the metadata

Does that imply adding a field in the `struct Meta`, and bumping `META_VERSION` ?

**Comment by t-bltg** [comment]
@cjdoris, can you remove the workflows approval please ?
It's cumbersome and wasting everybody's time.

**Comment by cjdoris** [comment]
> Thanks,
> 
> > the environment written to needs to be included in the metadata
> 
> Does that imply adding a `conda_env` field in the `struct Meta`, and bumping `META_VERSION` ?

Yes that's right. There should be a special value (the empty string?) for the default project location, in case the user moves the project.

I'll take a closer look at this in the new year, thank you.

**Comment by t-bltg** [comment]
> I'll take a closer look at this in the new year, thank you.

Thanks, have some nice afk holidays !

> Uh how? I only get a button to run the workflows (which I've just done).

Settings (the wheel) -> Actions (left side bar) -> General -> Change "Require approval for first-time contributors" to "Require approval for first-time contributors who are new to Github".

**Comment by cjdoris** [comment]
As you've gathered from Discourse I haven't forgotten about this PR! I'm just struggling to find enough laptop time to take a proper look.

Any thoughts on those CI failures?

One observation is you need to reset `STATE.conda_env` in `test/setup.jl`.

**Comment by t-bltg** [comment]
> One observation is you need to reset STATE.conda_env in test/setup.jl.

Thanks, I've changed that.

Can you apply the changes suggested in https://github.com/cjdoris/CondaPkg.jl/pull/53#issuecomment-1361749851 ?

CI is waiting again for approval: I can't thus move forward on this PR. Let's not waste time.

**Comment by t-bltg** [comment]
@cjdoris, CI is green now, so I think that this PR is ready for comments / review.

File: src/resolve.jl [hunk 1]
```diff
@@ -490,27 +497,28 @@ function resolve(; force::Bool=false, io::IO=stderr, interactive::Bool=false, dr
...
-            end
+            changed || _log(io, "Dependencies already up to date")
         else
             # the state is too dirty, recreate the conda environment from scratch
             dry_run && return
             # remove environment
             mkpath(meta_dir)
-            if isdir(conda_env)
+            create = true
+            isdir(conda_env) && if shared
```
Comment (cjdoris) [inline src/resolve.jl:507]: Combining `&&` and `if` in a single expression is kinda confusing - please just make it two `if`s.

**Review by cjdoris — CHANGES_REQUESTED** [review]
The new functionality looks great. I've had a close look at the code and it looks sound, plus I've played around with it and it seems to work.

Thanks for adding tests too, they look good.

File: src/resolve.jl [hunk 2]
```diff
@@ -435,13 +448,13 @@ function resolve(; force::Bool=false, io::IO=stderr, interactive::Bool=false, dr
...
-        if (meta === nothing) || (back == :Current)
+        if (meta === nothing) || (back === :Current)
             removed_pkgs = String[]
             changed_pkgs = String[]
-            added_pkgs = unique!(String[x.name for x in specs])
+            added_pkgs = unique!(map(x -> x.name, specs))
             removed_pip_pkgs = String[]
             changed_pip_pkgs = String[]
-            added_pip_pkgs = unique!(String[x.name for x in pip_specs])
+            added_pip_pkgs = unique!(map(x -> x.name, pip_specs))
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply src/resolve.jl]: You're right, restored.

File: src/resolve.jl [hunk 3]
```diff
@@ -386,18 +393,24 @@ function resolve(; force::Bool=false, io::IO=stderr, interactive::Bool=false, dr
     end
     STATE.resolved = false
     STATE.load_path = load_path
-    if back == :Current
+    if back === :Current
         # use a pre-existing conda environment
-        conda_env = get(ENV, "CONDA_PREFIX", "")
-        if conda_env == ""
+        if (conda_env = get(ENV, "CONDA_PREFIX", nothing)) === nothing
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply src/resolve.jl]: Restored.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -36,26 +42,17 @@ jobs:
             arch: x64
             backend: 'Null'
     steps:
-      - uses: actions/checkout@v2
-      - uses: julia-actions/setup-julia@v1
+      - uses: actions/checkout@v3
+      - uses: julia-actions/setup-julia@latest
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply .github/workflows/tests.yml]: Changed to fixed versions.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -7,6 +8,11 @@ on:
     branches:
       - main
     tags: '*'
+
+concurrency:
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply .github/workflows/tests.yml:12]: This allows to cancel the previous jobs in a PR. It avoids filling up the jobs queue when multiple commits are pushed in a small time interval, or if a branch is force pushed a lot.
It doesn't have effects on branches such as `master` or `main` (restricted to PRs), and the cancel is restricted to a unique PR.

File: src/resolve.jl [hunk 6]
```diff
@@ -365,14 +365,21 @@ function offline()
...
 function resolve(; force::Bool=false, io::IO=stderr, interactive::Bool=false, dry_run::Bool=false)
     # if frozen, do nothing
-    if STATE.frozen
-        return
-    end
+    STATE.frozen && return
     # if backend is Null, assume resolved
     back = backend()
-    if back == :Null
+    if back === :Null
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply src/resolve.jl:382]: Using `==` instead of `===` can cause some invalidations (e.g. https://github.com/JuliaPlots/Plots.jl/pull/4622). Furthermore, it is a usual convention to use triples for symbol comparisons.

File: src/resolve.jl [hunk 7]
```diff
@@ -386,18 +393,24 @@ function resolve(; force::Bool=false, io::IO=stderr, interactive::Bool=false, dr
...
         # find the topmost env in the load_path which depends on CondaPkg
         top_env = _resolve_top_env(load_path)
         STATE.meta_dir = meta_dir = joinpath(top_env, ".CondaPkg")
-        conda_env = joinpath(meta_dir, "env")
+        STATE.shared = shared = haskey(ENV, "JULIA_CONDAPKG_ENV")
+        STATE.conda_env = conda_env = if shared
+            ENV["JULIA_CONDAPKG_ENV"]
+        else
+            joinpath(meta_dir, "env")
+        end
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply src/resolve.jl:415]: We agree to disagree on the code style (I find it less readable), however, this is not my repo so I won't argue about this. I've added the suggestion.

File: src/resolve.jl (hunk 1 above)
Comment (t-bltg) [reply src/resolve.jl]: As you wish.

File: src/deps.jl [hunk 8]
```diff
@@ -308,15 +308,12 @@ function _add_temp!(dict, file, pkg)
 end
 
 function rm(pkgs::AbstractVector; resolve=true, file=cur_deps_file(), temp=false)
```
[the remark this answers is not in the document]
Comment (t-bltg) [reply src/deps.jl:310]: I've tried to revert it, but it fails the added tests (https://github.com/cjdoris/CondaPkg.jl/actions/runs/3974057027), so I've left this.

**Comment by t-bltg** [comment]
Thanks for the initial review.
I think I've reverted unrelated stylistic changes and addressed all your comments.

File: src/deps.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (t-bltg) [reply src/deps.jl]: I've adjusted the tests in https://github.com/cjdoris/CondaPkg.jl/pull/53/commits/a32a6f126fcfe61efeb053f307e0348f8e4d4b9f, and remove the `rm` early return.
It should be working now (got a bit annoyed by spurious windows ci failures).

File: src/resolve.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjdoris) [reply src/resolve.jl:382]: TIL, it even says to use `===` in the Julia docs.

File: .github/workflows/tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjdoris) [reply .github/workflows/tests.yml:12]: OK thanks!

**Review by cjdoris — COMMENTED** [review]
The code looks great now, thanks for your effort on this.

One thing that concerns me is that it's possible to end up with a Conda environment that is not consistent with your CondaPkg.toml. For example, in one project I specify `somepackage==1.0.0`, which is installed, then in another I specify `somepackage==2.0.0`, which is installed, then when I switch back to the first project, CondaPkg does not resolve again, so is still on 2.0.0. This is because the CondaPkg metadata is still stored per-project.

Some solutions to this:
1. Just document it as a current drawback of using a shared environment, leaving it up to the user to re-resolve when they change project (or ignore the problem, since dependencies are probably pretty loose anyway).
2. Don't allow `resolve()` to bail out early when in a shared environment. This will slow down time-to-first-resolve even when the environment is already OK.
3. Add a metadata file to the shared Conda environment specifying which Julia projects have installed stuff into it. Then CondaPkg can take dependencies from all of these projects. This could be a fair bit of work to implement.
4. Add a metadata file to the shared Conda environment specifying which Julia project was the last to resolve there. Then if you start Julia again in the same project, it can skip resolving. Downside is that it will do a full resolve whenever you swap projects.

The third option I think is the only one which would reliably always have the right dependencies, without any needless resolving.

I'd be happy to leave this as future work and merge this PR as-is, maybe adding a note to the documentation as in option 1.

**Comment by t-bltg** [comment]
Yes I first took the approach of writing a metadata file in the shared env (without considering one in a per-julia-version `.CondaPkg` directory), but resumed to an orthogonal direction after the first review in https://github.com/cjdoris/CondaPkg.jl/issues/52#issuecomment-1345606052.

So I think I'll probably document 1., and leave this enhancement for another PR.

Also, for my usage, having different specified conda packages versions never came up (kind of orthogonal to having a `shared` environment).

**Comment by t-bltg** [comment]
As a side note, the tests are hard to follow in CI, and IMO it would be useful to print the test name to split the `CondaPkg / MicroMamba` stdout output into per-test sections.

**Comment by t-bltg** [comment]
Hum, I fail to understand why `windows` CI [spuriously fails](https://github.com/cjdoris/CondaPkg.jl/actions/runs/3980417876/jobs/6823478550#step:6:1385).

**Comment by t-bltg** [comment]
I'm pretty sure we hit an incorrect fast path [here](https://github.com/cjdoris/CondaPkg.jl/actions/runs/3980545296/jobs/6823698425#step:6:1453).

File: src/resolve.jl [hunk 9]
```diff
@@ -242,6 +242,7 @@ function _resolve_conda_remove_all(io, conda_env)
 end
 
 function _resolve_conda_install(io, conda_env, specs, channels; create=false)
+    (length(specs) == 0 && !create) && return  # installing 0 packages is invalid
```
Comment (t-bltg) [inline src/resolve.jl:245]: This avoids https://github.com/cjdoris/CondaPkg.jl/actions/runs/3980733287/jobs/6824005615.

File: test/setup.jl [hunk 10]
```diff
@@ -18,3 +18,9 @@ CondaPkg.STATE.resolved = false
 CondaPkg.STATE.load_path = String[]
 CondaPkg.STATE.meta_dir = ""
 CondaPkg.STATE.frozen = false
+CondaPkg.STATE.conda_env = ""
+CondaPkg.STATE.shared = false
+
+empty!(CondaPkg.TEMP_PKGS)
```
Comment (t-bltg) [inline test/setup.jl:24]: Tests should not depend on a global state.

File: test/main.jl [hunk 11]
```diff
@@ -107,15 +108,39 @@ end
...
-    if testgc
-        CondaPkg.gc()
+    dn = string(tempname(), backend, Sys.KERNEL, VERSION)
+    @show dn
+    isnull || withenv("JULIA_CONDAPKG_ENV" => dn) do
+        # create empty env
+        CondaPkg.resolve()
+        @test !occursin("ca-certificates", status())
+        # add a package to specs and install it
+        CondaPkg.add("ca-certificates"; interactive=true, force=true)  # force: spurious windows failures
```
Comment (t-bltg) [inline test/main.jl:123]: Using `add` followed by `resolve(; force=true)` seems to fix the `windows` sporadic failures.

Comment (cjdoris) [reply test/main.jl:123]: Something to investigate separately, thanks.

**Comment by cjdoris** [comment]
This is great, thanks for all your work!
